<a href="https://colab.research.google.com/github/ibrahin12/Coursera_Capstone/blob/master/Buffalo_Final_Capstone.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 1. Introduction

##1a. Background:
 Buffalo, NY is the second largest city in New York. Millions of people travel to Niagara Falls yearly to experience the scenery of the falls. Tourists can experience the falls up close in a take a boat ride. Niagara Falls provides a tourist attraction which is beneficial for businesses like restaurants and hotels since many will travel from across different states in the US to experience the falls.

In addition, Buffalo is a city of sports; it is home to the Buffalo Bills and Sabres. Sports fanatics are drawn to the city to watch these sports games whether it be in a stadium or a restaurant. This provides competition for businesses to offer wings, pizza, and drinks to appeal to the sports crowd. Buffalo Wings are known to have originated in Buffalo and are a popular food among sports fans. The first wings were believed to have been from a family owned restaurant known as Anchor Bar. Thus, this provides competition for nearby restaurants to provide food such as wings that will draw in consumers. 

##1b. Problem:
There are many factors that need to be taken into consideration when determining where the best location is to open up a wing restaurant in Buffalo. Data that would be contributing to determine where to place a Buffalo wings restaurant is proximity to Niagara Falls. The proximity to Niagara Falls is an important location to consider because it is such a high tourist spot. In addition, proximity to other nearby restaurants or restaurants that are similar is another factor that needs to be considered. If someone wants to watch sports and grab quick food like wings they are likely to think of where other wing locations are and go there. Therefore, opening up a business that is nearby other similar businesses is a factor that needs to be considered. Thus, examining areas with high population densities and looking at venues under the list of postal codes obtained will help in deciding where to open up a Buffalo Wings restaurant. 

##1c. Interest: 
Tourists who are in Buffalo to see Niagara Falls would be interested in a nearby location to grab food and drinks that is quick and convenient and allows for them to experience traditional Buffalo wings. Furthermore, Buffalo is home to the Buffalo Bills and Buffalo Sabres, therefore, sports fanatics would be drawn to a place where they can have food and drinks and be able to watch the sports games. 

# 2.  Description of Data/How It Will Be Used

## 2a. K-Means Clustering:
Using an online html dataset https://data.mongabay.com/igapo/zip_codes/metropolitan-areas/metro-zip/Buffalo-Niagara%20Falls%20(NY)1.html which contains postal codes of Buffalo- Niagara Falls Metro Area, allows for clustering the dataset, wrangling the dataset, and reading the dataset into a pandas dataframe. The venue information obtained from Foursquare API will allow neighborhoods to be clustered based on similar venues in order to see where the best location would be for the restaurant. 

## 2b. Foursquare API:

Location data is an important factor to look at and is helpful in describing factors such as geographical locations. Given the location, we are then able to see if there are restaurants nearby. For determining where to open the Buffalo Wings restaurant, FourSquare API will be used to search for the latitudes and longitudes of the postal codes obtained from the K-Means clustering portion. This is done in order to obtain information about nearby venues, the category of the venues as well. If there was more than one town it was listed under the shared postal code so that each postal code appeared once. This provided the output of 110 different zipcodes that needed to be examined to obtain the venues nearby. Therefore, 110 is determined to be the limit of venues.

## 2c. Folium Markers Map:
Folium was used to create a map that has markers to indicate the population density of each neighborhood. If there are more restaurants that provide wings nearby, then that might be a good location to open up a business since consumers are familiar with the older established restaurants and are likely to go in that vicinity. Data was obtained from https://statisticalatlas.com/metro-area/New-York/Buffalo/Population. This website provided the population density for each subdivision. I then found the subdivisions latitudes and longitudes on Google Maps and used the population density obtained from statistical atlas to create a csv file. Once the file was made a map was created with a marker to indicate the population density in each subdivision. 


In [0]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests
pd.set_option("display.max_rows",999)

In [0]:
# getting data from internet
buffalo_link='https://data.mongabay.com/igapo/zip_codes/metropolitan-areas/metro-zip/Buffalo-Niagara%20Falls%20(NY)1.html'
raw_buffalo_page= requests.get(buffalo_link).text

# using beautiful soup to parse the HTML/XML codes.
soup = BeautifulSoup(raw_buffalo_page)
#print(soup.prettify())

Accessing the dataset in the table

In [0]:
table = soup.find('table', {'class':'boldtable'}) 

df_list= []
# extracting a clean form of the table
for tr_cell in soup.find_all('tr'):
    for td_cell in tr_cell.find_all('td'):
        df_list.append(td_cell.text.strip())

df_list[123:-2]

df_list=df_list[123:-2]

In [0]:
zipTown = df_list[0::6]
zipTown = zipTown[232:]
areaCode = df_list[1::6]
areaCode = areaCode[232:]
county = df_list[2::6]
county = county[232:]
state = df_list[3::6]
state = state[232:]
metroArea = df_list[4::6]
metroArea = metroArea[232:]
etcList = df_list[5::6]
etcList = etcList[232:]

In [0]:
zipCodes = []
town = []
for i in zipTown:
    zips = i[0:5]
    towns = i[6:]
    zipCodes.append(zips)
    town.append(towns)

In [0]:
metroArea2 = []
for j in metroArea:
    metar = j[0:25]
    metroArea2.append(metar)

In [0]:
df_buffalo = pd.DataFrame()

In [0]:
df_buffalo['Zip Codes'] = zipCodes
df_buffalo['Town'] = town
df_buffalo['Area Code'] = areaCode
df_buffalo['County'] = county
df_buffalo['State'] = state
df_buffalo['Metropolitan Area'] = metroArea2

In [0]:
df_buffalo=df_buffalo[:197]

In [0]:
df_buffalo.tail()

Zip Codes           Town  ...          State          Metropolitan Area
192     14304  Niagara Falls  ...  New York - NY  Buffalo-Niagara Falls, NY
193     14304    North Falls  ...  New York - NY  Buffalo-Niagara Falls, NY
194     14304     Wheatfield  ...  New York - NY  Buffalo-Niagara Falls, NY
195     14305  Niagara Falls  ...  New York - NY  Buffalo-Niagara Falls, NY
196     14305    North Falls  ...  New York - NY  Buffalo-Niagara Falls, NY

[5 rows x 6 columns]

In [0]:
df_update=df_buffalo.groupby('Zip Codes')['Town'].apply(lambda x: "%s" % ', '.join(x))

In [0]:
df_update

Zip Codes
14001                                      Akron, Newstead
14004                                      Alden, Townline
14006                                               Angola
14008                                             Appleton
14010                                        Athol Springs
14012                                               Barker
14025                                               Boston
14026                                         Bowmansville
14027                                                Brant
14028                                                 Burt
14030                                              Chaffee
14031                                             Clarence
14032                                      Clarence Center
14033                                               Colden
14034                                     Collins, Helmuth
14035                                       Collins Center
14038                                         

 Narrowing down the data to just zipcodes and towns 

In [0]:
df_update=df_update.reset_index()

In [0]:
df_update

Zip Codes                                               Town
0       14001                                    Akron, Newstead
1       14004                                    Alden, Townline
2       14006                                             Angola
3       14008                                           Appleton
4       14010                                      Athol Springs
5       14012                                             Barker
6       14025                                             Boston
7       14026                                       Bowmansville
8       14027                                              Brant
9       14028                                               Burt
10      14030                                            Chaffee
11      14031                                           Clarence
12      14032                                    Clarence Center
13      14033                                             Colden
14      14034                                   Collins, Helmuth
15      14035                                     Collins Center
16      14038                                         Crittenden
17      14043                      Cheektowaga, Depew, Lancaster
18      14047                                              Derby
19      14051                  Amherst, East Amherst, Swormville
20      14052                                        East Aurora
21      14055                                       East Concord
22      14057                                               Eden
23      14059                                               Elma
24      14061                                            Farnham
25      14067                                            Gasport
26      14068                                 Amherst, Getzville
27      14069                                           Glenwood
28      14072                                       Grand Island
29      14075                                            Hamburg
30      14080                                            Holland
31      14085                                Lake View, Lakeview
32      14086                                          Lancaster
33      14091                                            Lawtons
34      14092            Lewiston, Stela Niagara, Stella Niagara
35      14094                                Lockport, Pendleton
36      14095                                           Lockport
37      14102                                            Marilla
38      14105                       Middleport, Royalton, Shelby
39      14107                                         Model City
40      14108                                            Newfane
41      14109                                 Niagara University
42      14110                                       North Boston
43      14111                                      North Collins
44      14112                                        North Evans
45      14120  No Tonawanda, North Tonawanda, Pendleton, Whea...
46      14126                                             Olcott
47      14127                                       Orchard Park
48      14131                                        Ransomville
49      14132                                 Pendleton, Sanborn
50      14134                                           Sardinia
51      14139                                        South Wales
52      14140                          Spring Brook, Springbrook
53      14141                                        Springville
54      14144                      Stela Niagara, Stella Niagara
55      14150                                          Tonawanda
56      14151                                          Tonawanda
57      14169                                       Wales Center
58      14170                                         West Falls
59      14172                                             Wilson
60      14174                          

# Using Folium 

In [0]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize


import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Folium installed
Libraries imported.


# Foursquare API

FourSquare API will be used to search for the latitudes and longitudes of the postal codes obtained from the csv file that was created with the latitude and longitudes and the zip codes of areas in Buffalo. 

In [0]:
CLIENT_ID = '5WQGMO50PQ2AHHK2VK1GOVEOXIJU2Q1RMLMXWGQEUZ4XEKYN' # your Foursquare ID
CLIENT_SECRET = 'MFA4JYZGATRGFGFESWCI0FD4ULYBQMFCKV0BJPFO4OMQT5EV' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 5WQGMO50PQ2AHHK2VK1GOVEOXIJU2Q1RMLMXWGQEUZ4XEKYN
CLIENT_SECRET:MFA4JYZGATRGFGFESWCI0FD4ULYBQMFCKV0BJPFO4OMQT5EV


In [0]:
geolocator=Nominatim()

/usr/local/lib/python3.6/dist-packages/geopy/geocoders/osm.py:143: UserWarning: Using Nominatim with the default "geopy/1.17.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  UserWarning


In [0]:
df_update['Zip Codes'].tolist()

['14001',
 '14004',
 '14006',
 '14008',
 '14010',
 '14012',
 '14025',
 '14026',
 '14027',
 '14028',
 '14030',
 '14031',
 '14032',
 '14033',
 '14034',
 '14035',
 '14038',
 '14043',
 '14047',
 '14051',
 '14052',
 '14055',
 '14057',
 '14059',
 '14061',
 '14067',
 '14068',
 '14069',
 '14072',
 '14075',
 '14080',
 '14085',
 '14086',
 '14091',
 '14092',
 '14094',
 '14095',
 '14102',
 '14105',
 '14107',
 '14108',
 '14109',
 '14110',
 '14111',
 '14112',
 '14120',
 '14126',
 '14127',
 '14131',
 '14132',
 '14134',
 '14139',
 '14140',
 '14141',
 '14144',
 '14150',
 '14151',
 '14169',
 '14170',
 '14172',
 '14174',
 '14201',
 '14202',
 '14203',
 '14204',
 '14205',
 '14206',
 '14207',
 '14208',
 '14209',
 '14210',
 '14211',
 '14212',
 '14213',
 '14214',
 '14215',
 '14216',
 '14217',
 '14218',
 '14219',
 '14220',
 '14221',
 '14222',
 '14223',
 '14224',
 '14225',
 '14226',
 '14227',
 '14228',
 '14231',
 '14233',
 '14240',
 '14241',
 '14260',
 '14261',
 '14263',
 '14264',
 '14265',
 '14267',
 '14269',


 Created csv of latitudes and longitudes of Buffalo Zipcodes to Use as input for Foursquare API

In [0]:
buff_coordinates=pd.read_csv('/content/drive/My Drive/Buffalo Zip Code Coordinates.csv')

In [0]:
buff_coordinates.head()


Zipcode  Latitude   Longitude
0    14001   43.02336  -78.49944
1    14004   42.89995  -78.46411
2    14006   42.64108  -79.05541
3    14008   43.32711  -78.64915
4    14010   42.71599  -78.83124

In [0]:
buff_coordinates.columns

Index(['Zipcode', 'Latitude ', 'Longitude'], dtype='object')

# Searching For Nearby Venues Using Foursquare API

In [0]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        #print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET,
            VERSION,
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [0]:

buffalo_venues = getNearbyVenues(names=buff_coordinates['Zipcode'],
                                   latitudes=buff_coordinates['Latitude '],
                                   longitudes=buff_coordinates['Longitude']
                                  )

In [0]:
buffalo_venues

Neighborhood  ...                   Venue Category
0           14001  ...              American Restaurant
1           14001  ...                   Sandwich Place
2           14001  ...                              Pub
3           14001  ...               Italian Restaurant
4           14001  ...                         Pharmacy
5           14001  ...                     Optical Shop
6           14001  ...               Chinese Restaurant
7           14001  ...                     Liquor Store
8           14001  ...                      Pizza Place
9           14001  ...                    Deli / Bodega
10          14006  ...                   Sandwich Place
11          14006  ...                Convenience Store
12          14006  ...                     Burger Joint
13          14010  ...                             Café
14          14010  ...                    Movie Theater
15          14010  ...                        Gastropub
16          14010  ...                              Pub
17          14010  ...                      Coffee Shop
18          14010  ...                   Ice Cream Shop
19          14010  ...                      Pizza Place
20          14010  ...                     Burger Joint
21          14010  ...                             Café
22          14010  ...               Mexican Restaurant
23          14010  ...                    Bowling Alley
24          14010  ...                              Bar
25          14010  ...                      Pizza Place
26          14010  ...                   Farmers Market
27          14010  ...                      Pizza Place
28          14010  ...              American Restaurant
29          14010  ...                   Discount Store
30          14010  ...                     Intersection
31          14010  ...                French Restaurant
32          14010  ...                         Pharmacy
33          14010  ...               Athletics & Sports
34          14026  ...       Construction & Landscaping
35          14026  ...                             Food
36          14027  ...       Construction & Landscaping
37          14027  ...                            Plaza
38          14027  ...                     Food Service
39          14031  ...               Mexican Restaurant
40          14031  ...                           Bakery
41          14031  ...                       Steakhouse
42          14031  ...                      Flower Shop
43          14031  ...                              Gym
44          14031  ...                     Cupcake Shop
45          14032  ...                     Intersection
46          14033  ...              American Restaurant
47          14033  ...                Convenience Store
48          14034  ...                      Pizza Place
49          14034  ...       Construction & Landscaping
50          14035  ...                 Recording Studio
51          14035  ...                             Park
52          14038  ...                              Bar
53          14038  ...                     Intersection
54          14043  ...                       Sports Bar
55          14043  ...                      Wings Joint
56          14043  ...              Rental Car Location
57          14043  ...                         Pharmacy
58          14043  ...                             Food
59          14043  ...                     Intersection
60          14043  ...                    Deli / Bodega
61          14043  ...                        Speakeasy
62          14043  ...             Fast Food Restaurant
63          14043  ...                     Intersection
64          14043  ...                Recreation Center
65          14043  ...                        BBQ Joint
66          14043  ...                              Bar
67          14043  ...                       Smoke Shop
68          14047  ...                              Gym
69          14047  ...                         Pharmacy
70          14047  ...       

# Analyzing Each Neighborhood (Zipcode)

In [0]:
# one hot encoding
buffalo_onehot = pd.get_dummies(buffalo_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
buffalo_onehot['Zipcode'] = buffalo_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [buffalo_onehot.columns[-1]] + list(buffalo_onehot.columns[:-1])
buffalo_onehot = buffalo_onehot[fixed_columns]

buffalo_onehot.head()

Zipcode  ATM  Accessories Store  ...  Wings Joint  Women's Store  Yoga Studio
0    14001    0                  0  ...            0              0            0
1    14001    0                  0  ...            0              0            0
2    14001    0                  0  ...            0              0            0
3    14001    0                  0  ...            0              0            0
4    14001    0                  0  ...            0              0            0

[5 rows x 217 columns]

In [0]:
buffalo_onehot.shape

(936, 217)

# Grouping each neighborhood to obtain the average frequency of each category

In [0]:
buffalo_grouped = buffalo_onehot.groupby('Zipcode').mean().reset_index()
buffalo_grouped

Zipcode       ATM  ...  Women's Store  Yoga Studio
0     14001  0.000000  ...       0.000000     0.000000
1     14006  0.000000  ...       0.000000     0.000000
2     14010  0.000000  ...       0.000000     0.000000
3     14026  0.000000  ...       0.000000     0.000000
4     14027  0.000000  ...       0.000000     0.000000
5     14031  0.000000  ...       0.000000     0.000000
6     14032  0.000000  ...       0.000000     0.000000
7     14033  0.000000  ...       0.000000     0.000000
8     14034  0.000000  ...       0.000000     0.000000
9     14035  0.000000  ...       0.000000     0.000000
10    14038  0.000000  ...       0.000000     0.000000
11    14043  0.000000  ...       0.000000     0.000000
12    14047  0.000000  ...       0.000000     0.000000
13    14051  0.000000  ...       0.000000     0.000000
14    14052  0.000000  ...       0.000000     0.000000
15    14057  0.000000  ...       0.000000     0.000000
16    14059  0.000000  ...       0.000000     0.000000
17    14061  0.000000  ...       0.000000     0.000000
18    14067  0.000000  ...       0.000000     0.000000
19    14068  0.000000  ...       0.000000     0.000000
20    14069  0.000000  ...       0.000000     0.000000
21    14072  0.000000  ...       0.000000     0.000000
22    14075  0.000000  ...       0.000000     0.000000
23    14080  0.000000  ...       0.000000     0.000000
24    14086  0.000000  ...       0.000000     0.000000
25    14092  0.000000  ...       0.000000     0.000000
26    14095  0.000000  ...       0.000000     0.000000
27    14102  0.000000  ...       0.000000     0.000000
28    14105  0.000000  ...       0.000000     0.000000
29    14107  0.000000  ...       0.000000     0.000000
30    14108  0.000000  ...       0.000000     0.000000
31    14109  0.000000  ...       0.000000     0.000000
32    14120  0.000000  ...       0.000000     0.000000
33    14126  0.000000  ...       0.000000     0.000000
34    14127  0.000000  ...       0.000000     0.000000
35    14131  0.000000  ...       0.000000     0.000000
36    14132  0.166667  ...       0.000000     0.000000
37    14134  0.000000  ...       0.000000     0.000000
38    14139  0.000000  ...       0.000000     0.000000
39    14141  0.000000  ...       0.000000     0.000000
40    14144  0.000000  ...       0.000000     0.000000
41    14150  0.000000  ...       0.000000     0.000000
42    14169  0.000000  ...       0.000000     0.000000
43    14172  0.000000  ...       0.000000     0.000000
44    14174  0.000000  ...       0.000000     0.000000
45    14201  0.000000  ...       0.000000     0.000000
46    14202  0.000000  ...       0.000000     0.000000
47    14203  0.000000  ...       0.000000     0.000000
48    14204  0.000000  ...       0.000000     0.000000
49    14205  0.000000  ...       0.000000     0.000000
50    14206  0.000000  ...       0.000000     0.000000
51    14207  0.000000  ...       0.000000     0.000000
52    14208  0.000000  ...       0.000000     0.000000
53    14209  0.000000  ...       0.000000     0.000000
54    14210  0.000000  ...       0.000000     0.000000
55    14211  0.000000  ...       0.000000     0.000000
56    14212  0.000000  ...       0.000000     0.000000
57    14213  0.000000  ...       0.058824     0.058824
58    14214  0.000000  ...       0.000000     0.000000
59    14215  0.000000  ...       0.000000     0.000000
60    14216  0.000000  ...       0.000000     0.000000
61    14217  0.000000  ...       0.000000     0.000000
62    14218  0.000000  ...       0.000000     0.000000
63    14219  0.000000  ...       0.000000     0.000000
64    14220  0.000000  ...       0.000000     0.000000
65    14221  0.000000  ...       0.000000     0.000000
66    14222  0.000000  ...       0.000000     0.000000
67    14223  0.000000  ...       0.000000     0.000000
68    14224  0.000000  ...       0.000000     0.000000
69    14225  0.000000  ...       0.000000     0.000000
70    14226  0.000000  ...       0.000000     0.000000
71    14227  0.000000  ...       0.000000   

In [0]:
buffalo_grouped.shape

(94, 217)

# Examining each neighborhood with their top 5 venues

In this portion the foursquare API venues will be assessed to determine the top venues in each zip code.

In [0]:
num_top_venues = 5

for hood in buffalo_grouped['Zipcode']:
    temp = buffalo_grouped[buffalo_grouped['Zipcode'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

                venue  freq
0       Deli / Bodega   0.1
1      Sandwich Place   0.1
2        Optical Shop   0.1
3  Chinese Restaurant   0.1
4        Liquor Store   0.1


               venue  freq
0       Burger Joint  0.33
1  Convenience Store  0.33
2     Sandwich Place  0.33
3                ATM  0.00
4               Park  0.00


           venue  freq
0    Pizza Place  0.14
1           Café  0.10
2   Intersection  0.05
3  Bowling Alley  0.05
4            Bar  0.05


                        venue  freq
0  Construction & Landscaping   0.5
1                        Food   0.5
2                         ATM   0.0
3                        Park   0.0
4                   Multiplex   0.0


                           venue  freq
0                          Plaza  0.33
1     Construction & Landscaping  0.33
2                   Food Service  0.33
3                            ATM  0.00
4  Paper / Office Supplies Store  0.00


                venue  freq
0          Steakhouse  0.17
1        Cupcake

In [0]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

# Creating A New Dataframe With The Top 10 Venues Of Each Neighborhood

In [0]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Zipcode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Zipcode'] = buffalo_grouped['Zipcode']

for ind in np.arange(buffalo_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(buffalo_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Zipcode  ... 10th Most Common Venue
0    14001  ...    American Restaurant
1    14006  ...      Food & Drink Shop
2    14010  ...                    Bar
3    14026  ...             Food Court
4    14027  ...      Food & Drink Shop

[5 rows x 11 columns]

# Clustering Venues in Neighborhoods
In this section venues were clustered under a common group (zip code) 

In [0]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

In [0]:
# set number of clusters
kclusters = 5

buffalo_grouped_clustering = buffalo_grouped.drop('Zipcode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(buffalo_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 2, 1, 1, 1, 1, 0, 2, 1, 2], dtype=int32)

In [0]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

buffalo_merged = buff_coordinates
buffalo_merged = buffalo_merged.join(neighborhoods_venues_sorted.set_index('Zipcode'), on='Zipcode')
buffalo_merged.head()

Zipcode  Latitude   ...  9th Most Common Venue  10th Most Common Venue
0    14001   43.02336  ...                    Pub     American Restaurant
1    14004   42.89995  ...                    NaN                     NaN
2    14006   42.64108  ...             Food Court       Food & Drink Shop
3    14008   43.32711  ...                    NaN                     NaN
4    14010   42.71599  ...              Gastropub                     Bar

[5 rows x 14 columns]

In [0]:
buffalo_merged

Zipcode  ...           10th Most Common Venue
0      14001  ...              American Restaurant
1      14004  ...                              NaN
2      14006  ...                Food & Drink Shop
3      14008  ...                              NaN
4      14010  ...                              Bar
5      14012  ...                              NaN
6      14025  ...                              NaN
7      14026  ...                       Food Court
8      14027  ...                Food & Drink Shop
9      14028  ...                              NaN
10     14030  ...                              NaN
11     14031  ...                           Forest
12     14032  ...                             Food
13     14033  ...                       Food Court
14     14034  ...                Food & Drink Shop
15     14035  ...                             Food
16     14038  ...                       Food Court
17     14043  ...                Recreation Center
18     14047  ...                Fish & Chips Shop
19     14051  ...                   Clothing Store
20     14052  ...                             Café
21     14055  ...                              NaN
22     14057  ...                       Food Truck
23     14059  ...                       Food Court
24     14061  ...                Food & Drink Shop
25     14067  ...                       Food Court
26     14068  ...                Food & Drink Shop
27     14069  ...                Food & Drink Shop
28     14072  ...                      Music Store
29     14075  ...       Construction & Landscaping
30     14080  ...                       Food Court
31     14085  ...                              NaN
32     14086  ...                       Food Court
33     14091  ...                              NaN
34     14092  ...                       Food Court
35     14094  ...                              NaN
36     14095  ...                    Tour Provider
37     14102  ...                       Food Court
38     14105  ...                       Food Truck
39     14107  ...                     Food Service
40     14108  ...                     Food Service
41     14109  ...                       Food Court
42     14110  ...                              NaN
43     14111  ...                              NaN
44     14112  ...                              NaN
45     14120  ...                Food & Drink Shop
46     14126  ...                      Flower Shop
47     14127  ...              American Restaurant
48     14131  ...                      Flower Shop
49     14132  ...                       Food Truck
50     14134  ...                Food & Drink Shop
51     14139  ...                       Food Court
52     14140  ...                              NaN
53     14141  ...                     Food Service
54     14144  ...                Food & Drink Shop
55     14150  ...           Furniture / Home Store
56     14151  ...                              NaN
57     14169  ...                Food & Drink Shop
58     14170  ...                              NaN
59     14172  ...                       Food Court
60     14174  ...                             Food
61     14201  ...                   Discount Store
62     14202  ...                         Beer Bar
63     14203  ...                          Theater
64     14204  ...                Food & Drink Shop
65     14205  ...                        BBQ Joint
66     14206  ...                Convenience Store
67     14207  ...                       Food Court
68     14208  ...                       Food Truck
69     14209  ...                Food & Drink Shop
70     14210  ...                Food & Drink Shop
71     14211  ...                       Food Truck
72     14212  ...                     Food Service
73     14213  ...                           Bakery
74     14214  ...                     Food Service
75     14215  ...                      Dry Cleaner
76     14216  ...                   Sandwich Place
77     14217  ...          

Removing 'NaN' venue zip codes from the dataset

In [0]:
#eliminating postal codes with no values
buffalo_merged.dropna(axis=0,inplace=True)

In [0]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [0]:
buffalo_merged= buffalo_merged.astype({'Cluster Labels':'int32'})

In [0]:
# create map
map_clusters = folium.Map(location=[42.8864, -78.8784], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(buffalo_merged['Latitude '], buffalo_merged['Longitude'], buffalo_merged['Zipcode'],buffalo_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Creating Folium Markers Map of Population Densities In Buffalo Subdivisions

In [0]:
import numpy as np  
import pandas as pd 

In [0]:
import folium

print('Folium installed and imported!')

Folium installed and imported!


In [0]:
# define the world map centered around Buffalo with a low zoom level
world_map = folium.Map(location=[42.8864, -78.8784], zoom_start=4)

# display world map
world_map

In [0]:
# define the world map centered around Buffalo with a low zoom level
world_map = folium.Map(location=[42.8864, -78.8784], zoom_start=12)

# display world map
world_map

In [0]:
world_geo = r'world_countries.json' # geojson file

# create a plain world map
world_map = folium.Map(location=[0, 0], zoom_start=2, tiles='Mapbox Bright')

In [0]:
df_buffalo_map=pd.read_csv('/content/drive/My Drive/Buffalo population density -final.csv')

In [0]:
df_buffalo_map

Subdivision   Latitude  Longitude Density
0                        Buffalo  42.886400 -78.878400   6,413
1                      Tonawanda  43.020300 -78.880300   3,941
2              Town of Tonawanda  42.989100 -78.872200   3,904
3                North Tonawanda  43.038700 -78.864200   3,063
4                  Niagara Falls  43.096200 -79.037700   3,491
5            Town of Cheektowaga  42.907100 -78.754300   2,976
6                     Lackawanna  42.825600 -78.823400   2,738
7                       Lockport  43.170600 -78.690300   2,467
8                Town of Amherst  42.979000 -78.792300   2,343
9            Town of West Seneca  42.835898 -78.753890   2,116
10               Town of Hamburg  42.715900 -78.829500   1,402
11             Town of Lancaster  42.901200 -78.670200   1,133
12               Town of Niagara  43.115300 -78.985400     863
13          Town of Orchard Park  42.767500 -78.744100     768
14          Town of Grand Island  43.018300 -79.013400     733
15            Town of Wheatfield  43.085500 -78.894000     655
16              Town of Clarence  42.976670 -78.591940     592
17              Town of Lockport  43.141500 -78.668200     451
18              Town of Lewiston  43.191900 -78.985400     432
19                 Town of Evans  42.663900 -78.996700     392
20                Town of Aurora  42.767700 -78.613300     380
21                  Town of Elma  42.821390 -78.636940     339
22                 Town of Alden  42.939600 -78.548700     298
23             Town of Pendleton  43.086940 -78.727780     242
24                Town of Boston  42.652800 -78.745200     224
25                Town of Porter  43.145900 -78.595300     201
26               Town of Marilla  42.840060 -78.555024     195
27                  Town of Eden  42.652100 -78.895500     193
28               Town of Newfane  43.282400 -78.707600     182
29              Town of Newstead  43.011700 -78.514200     171
30               Town of Cambria  43.163600 -78.819500     147
31               Town of Collins  42.495400 -78.919800     137
32               Town of Concord  42.337800 -78.420000     122
33                Town of Wilson  43.309800 -78.826100     120
34  Tuscarora Nation Reservation  43.158400 -78.941400     111
35              Town of Royalton  43.102400 -78.323900     108
36               Town of Holland  42.382800 -78.322800      94
37                Town of Colden  42.383900 -78.410500      92
38                 Town of Brant  42.588400 -79.015200      87
39                 Town of Wales  42.767900 -78.527300      85
40          Town of North Collin  42.354200 -78.562800      82
41              Town of Hartland  43.242100 -78.532800      77
42              Town of Somerset  43.347683 -78.555656      74
43       Cattaraugus Reservation  42.550900 -79.038600      74
44              Town of Sardinia  42.332300 -78.315200      55
45         Tonawanda Reservation  43.072300 -78.447200       2

In [0]:
df_buffalo_map.shape

(46, 4)

In [0]:
# instantiate a feature group for the densities in the dataframe
buffalo_map_features = folium.map.FeatureGroup()

# loop through the subdivisions and add each to the density feature group
for lat, lng in zip(df_buffalo_map['Latitude'], df_buffalo_map['Longitude']):
    buffalo_map_features.add_child(
        folium.vector_layers.CircleMarker(
            [lat, lng],
            radius=8, # define how big you want the circle markers to be
            color='yellow',
            fill=True,
            fill_color='blue',
            fill_opacity=0.6
        )
    )

# add pop-up text to each marker on the map
latitudes = list(df_buffalo_map.Latitude)
longitudes = list(df_buffalo_map.Longitude)
labels = list(df_buffalo_map['Density'])

for lat, lng, label in zip(latitudes, longitudes, labels):
    folium.Marker([lat, lng], popup=label).add_to(buffalo_map_features)
    print(label)    
    
# add densities to map
buffalo_map_features.add_child(buffalo_map_features)

6,413
3,941
3,904
3,063
3,491
2,976
2,738
2,467
2,343
2,116
1,402
1,133
863
768
733
655
592
451
432
392
380
339
298
242
224
201
195
193
182
171
147
137
122
120
111
108
94
92
87
85
82
77
74
74
55
2


In [0]:
# create map and display it
buffalo_map = folium.Map(location=[42.8864, -78.8784], zoom_start=9)

# loop through the 100 crimes and add each to the map
for lat, lng, label1 in zip(df_buffalo_map['Latitude'], df_buffalo_map['Longitude'],df_buffalo_map['Density']):
    folium.vector_layers.CircleMarker(
        [lat, lng],
        radius=8, # define how big you want the circle markers to be
        color='yellow',
        fill=True,
        popup=label1,
        fill_color='blue',
        fill_opacity=0.6
    ).add_to(buffalo_map)

# show map
buffalo_map

# Results
When examining the folium marker’s map with the population densities, the marker dot right above Buffalo, representing the subdivision of Buffalo, has the greatest population density of 6,413 (/mi^2). The next highest subdivision is Tonawanda which has a population density of 3,941 (/mi^2). When looking at the merged table the cluster label that appears the most is 1, which is reflected in the markers map. The greatest clustering markers are around Buffalo and there are no other sections with comparable marker clusters. 

#Discussion
Based on the results obtained from the the folium marker’s map it appears as if the largest population density is Buffalo, NY. This makes sense because Buffalo is the second largest city in New York. Therefore, it is expected to have the greatest population density. As seen through the information about the venues obtained from Foursquare API, some of the zip codes such as 14004 and 14008 were dropped from the venue table because there was no information about them. Therefore, the loss of these venues from the dataset lowered the number of postal codes in the data pool. However, since there were 110 zip codes that were being examined.
When determining the best location for the Buffalo Wings restaurant, a location with a cluster marker of 1 and a high population density would be the best location for a business. Opening up a restaurant in zip code 14273 is a location that falls near Buffalo, which has the largest population density, and it is under the cluster label 1. Looking at the venues in this location it appears as if the restaurants are similar to Buffalo Wings. The top 5 venues are music, bar, hotel, wine bar, and American restaurant. This location would be beneficial to locate a restaurant because hotels are the third common venues in the Buffalo, HSBC Atrium postal code area. Hotels being high up on the venue list indicates that it might be a location for tourists and where many might stay when they are visiting Buffalo. The other 5 common venues are sandwich place, steakhouse, multiplex, cocktail bar, and a BBQ joint. These venues indicate traditional American food, therefore creating a Buffalo Wings business would be beneficial to add to this complex because it would be surrounded by other common cuisines. Since bars are also common, it would be beneficial to add to this location because the restaurant that I would be pursuing in opening would allow for watching sports, having drinks, and eating wings among other traditional foods.  


#Conclusion
From the markers map of the subdivisions and the map of the venues, it can be concluded that there are many options for opening up a venue in close proximity. Any location that is around Buffalo would be a good place to open up a business, largely because it is the second largest city in New York and seems to be a place that is largely visited by tourists due to the fact that hotels are in the top 3 most common venues. There are other variables that would need to be considered when deciding where to place a restaurant such as the cost of opening up a business in Buffalo or in areas around Buffalo. More analysis would need to be done.  However, looking at the population densities and the most common venues that are in close proximity to the highly populated areas was a way to narrow down the search for opening a business. Assessing how close the postal code chosen is to Niagara Falls and the Buffalo Bills stadium were also two important factors to consider because both are driving factors for individuals coming to the city. 

# References
“Foursquare Developer.” Foursquare Developer, developer.foursquare.com/. 

Google Maps, Google, www.google.com/maps. 

“Latitude and Longitude Finder.” Lat Long Finder, www.latlong.net/. 

Mongabay. “Postal Codes for Buffalo-Niagara Falls (NY) Metro Area.” Mongabay.com, data.mongabay.com/igapo/zip_codes/metropolitan-areas/metro-zip/Buffalo-Niagara Falls (NY)1.html. 

“Population of the Buffalo Area, New York (Metro Area).” The Demographic Statistical Atlas of the United States - Statistical Atlas, statisticalatlas.com/metro-area/New-York/Buffalo/Population. 